### PR TITLE
feat(Server): add-on storage

### PR DESCRIPTION
**Pull request by BreadGenie**
Screenshot 2024-06-25 at 12 08 13 PM

https://github.com/frappe/press/assets/63963181/364d7644-befa-40ae-97b2-8fd5262628e3

- [x] Increasing disk size from UI
- [x] Billing
- [x] Patch to add subscription for existing users
- [x] Invoice view

### CODE REVIEW

File: press/press/doctype/server/server.py [hunk 1]
```diff
@@ -497,10 +507,63 @@ def _create_initial_plan_change(self, plan):
...
+			},
+			["name", "additional_storage"],
+			as_dict=True,
+		):
+			frappe.db.set_value(
+				"Subscription",
+				existing_subscription.name,
+				"additional_storage",
+				increment + int(existing_subscription.additional_storage),
+			)
```
Comment (shadrak98) [inline press/press/doctype/server/server.py:559]: @BreadGenie why not create a new Subscription record for new purchase?

The problem here in updating the existing subscription record is the plan remains the same. You are not updating the plan that means everytime someone buys additional storage he will be charged the same initial amount.
Correct me if my understanding is incorrect.

File: press/press/doctype/server/server.py [hunk 2]
```diff
@@ -1008,12 +1089,62 @@ def update_subscription(self):
...
+			},
+		):
+			add_on_storage_subscription = (
+				frappe.get_doc("Subscription", add_on_storage_subscription)
+				if add_on_storage_subscription
+				else None
+			)
+
+		if add_on_storage_subscription and add_on_storage_subscription.team != self.team:
+			add_on_storage_subscription.disable()
```
Comment (shadrak98) [inline press/press/doctype/server/server.py]: If we need `frappe.get_doc` for disable method then better to call it just before this line.

File: press/press/doctype/server/server.py (hunk 1 above)
Comment (BreadGenie) [reply press/press/doctype/server/server.py:559]: There'll be only one active storage plan at a time, and it'll only hold the per gb per month rate. The extra disk value is stored in subscription and the rate is computed with that as well

See https://github.com/frappe/press/blob/3a9f707eec9ff55ebda1e48db42ad51141bbd30f/press/press/doctype/subscription/subscription.py#L148-L149

**Review by shadrak98 — APPROVED** [review]

**Comment by rmehta** [comment]
@BreadGenie do show the increased amount per day as well. This is an important parameter to make this decision.

Also always get a UI review on such designs. (Happy to review)

**Comment by BreadGenie** [comment]
We do show the info in the plan section (per month though)

Screenshot 2024-06-29 at 11 42 54 AM

Maybe we can show the add-on plan details explicitly as well (without hover). Went this way because it looked cluttered when I tried to show all the details in that small box.

**Comment by rmehta** [comment]
Might be a good idea to show it when then are updating as well.

**Pull request merged**